docker/docker_client.go

```diff
@@ -42,8 +47,13 @@ type dockerClient struct {
 	transport       *http.Transport
 }
 
-// newDockerClient returns a new dockerClient instance for refHostname (a host a specified in the Docker image reference, not canonicalized to dockerRegistry)
-func newDockerClient(refHostname, certPath string, tlsVerify bool) (*dockerClient, error) {
+// NewClient returns a new Client instance for refHostname (a host a specified in the Docker image reference, not canonicalized to dockerRegistry)
+func NewClient(img, certPath string, tlsVerify bool) (Client, error) {
+	ref, _, err := parseDockerImageName(img)
```

Comment:
I’m thinking the client should be hidden from users of the library, and then AFAICS this problem disappears because the parsing would still happen in the `Image{,Source,Destination}`.

Separately, for policy namespacing I will probably need a transport-independent “parsed reference” (`types.ImageReference` interface), with external users first converting a string into an implementation of that interface (or perhaps constructing an implementation directly), and _then_ calling NewImage$something; so the parsing would happen separately and externally.